### ISSUE TITLE
Update ami_analyze step using ImPlaneIA algorithm

### ISSUE THREAD

**Issue by stscijgbot-jp** [spacetelescope/jwst#4935]
_Issue [JP-1449](https://jira.stsci.edu/browse/JP-1449) was created on JIRA by [David Grumm](https://jira.stsci.edu/secure/ViewProfile.jspa?name=grumm):_

Anand and Deepashri have delivered to us the ImPlaneIA algorithm to replace the ami_analyze step of the existing pipeline. The work to do this includes a review of the accompanying tutorials and notebooks (JP-1458), tests (JP-1459) , and documentation.

Several emails are included here for background.

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
Monday, May 4, 2020 at 1:06 PM

Hi Dave and Howard,

The ImPlaneIA algorithm to replace the ami_analyze step of the existing pipeline is available in the delivery branch of <https://github.com/anand0xff/ImPlaneIA/> repository.

Please refer to the file NOTES.md in this branch for delivery notes.

The driver code implaneia_driver.py also does software verification of the code.

Please let us know if you have any questions.

Thanks
Deepashri and Anand
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
Thursday, November 21, 2019 at 9:30 AM-
Hi Anand,

We just finished delivering and testing our CAL code for DMS Build 7.4. In a couple weeks we'll be starting to work on items that are already on the to-do list for B7.5. We have to complete our coding for B7.5 by the end of February 2020. Right now it's looking like B7.5 (or a potential B7.5.x patch for any serious bugs) may become the "flight" build - i.e. the version that gets frozen for use in the SOC during launch and commissioning. We will of course be continuing development up until and well beyond launch, such that upgrades will be available to anyone who wants to install the most recent versions on their own.

As of today, the to-do list for B7.5 development is already full with priority items from INS and the MO. So we really aren't in a position to accept any new work for enhancements like this any time soon. The baseline AMI code (LG+) that's currently in the pipeline will therefore need to be what's in the launch version and what therefore gets used in the DMS production environment. We could add this AMI upgrade to the list for B7.6 work, which we will start working on around April 2020 (after we get B7.5 tested and out the door). So I don't see any need for you to give us what you have now and have us start working on it, only to then make further updates later. We should just wait for a complete delivery of the whole enhanced version early next year (e.g. you aim to have it finished up by the end of March 2020 and then we can potentially start working on it after that).

Cheers,
Howard "SPP rocks" Bushouse

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
On 11/20/19, 9:11 PM, "Anand Sivaramakrishnan" <anand@stsci.edu> wrote:

Hi Howard and Dave

    We are in the following state...

    The new AMI reduction code (ImPlaneIA, aka LG++)  runs on MAST-header data cubes comprised of one or several integrations.  It runs with mirage headers grafted onto "data" without bad pixels.  It outputs text files of "observables" formatted the same way as LG+ and LG did.  It has some actual unit test code and some less stringent unit test code on the major changes from LG+.  I hope it is overall much cleaner (at least until it gets to "Calibrate").  The 'research grade' version Alex Greenbaum used a few moths ago works much better than the LG+ version (in the pipeline).

    In assembling old and new bits into this we are still perhaps not a dollar short, but more like 50c short in terms of scientific accuracy.  There appears to be at least one bug that causes accuracy to be suspect at a ~10-30% level.  It might be a 'dumb bug', not a fearsomely subtle one.  We are looking for it.

    The code is somewhat trustworthy (other than the dumb bug) until it gets to "Calibrate"ing data sets.  At that point it gets to be very involved, and needs cleanout and simplifying, just to be nimble on looking at COM and CAL data.  That nimbleness will be in how we feed it different data sets and peruse the results, not in any fancy internals of the basic extraction of observables.

    We had signed up for an end-of-November deadline on "delivery" to you. Where are you (temporally)  in terms of being eager to get the code and work on it?  Do you want to start with what we have and we keep going on improving parts to ImPlaneIA, and give you upgrades along the way, or do you want to wait for  "commissioning-ready" ImPlaneIA?  To be realistic, commissioning-ready ImPlaneIA will take two more months of focussed work by both Deepashri and me.  You might be able to absorb the changes from the current version two months later, without too much hassle, if you have our beta ImPlaneIA jwst-pipeline-ized by that time.  Maybe you'd like a code walkthrough soon, over about  two hours or so... we can certainly oblige.

    Let us know your druthers, and we'll talk to Steph about how to proceed.  If you want us to rewrite it in SPP or FORTRAN 2 please let us know (and we'll quit and become cosmologists).

    cheers
    Anand

**Comment by stscijgbot-jp** [comment]
_Comment by [Alicia Canipe](https://jira.stsci.edu/secure/ViewProfile.jspa?name=acanipe) on [JIRA](https://jira.stsci.edu/browse/JP-1449?focusedCommentId=464314#comment-464314):_

Just adding a note from Deepashri for work prioritization purposes, if needed:

 

> ImPlaneIA is critical for commissioning and/or science operations.
> 
> ImPlaneIA is a much improved version of the algorithm currently used by ami_analyze step of the pipeline. It is possible to use ImPlaneIA as a tool outside the pipeline to analyze data that is processed by Image1 and Image2 pipelines. However implementation in the pipeline would make the algorithm much more efficient for users, and enable commissioning data reduction using the pipeline implementation of ImPlaneIA.

**Comment by stscijgbot-jp** [comment]
_Comment by [David Grumm](https://jira.stsci.edu/secure/ViewProfile.jspa?name=grumm) on [JIRA](https://jira.stsci.edu/browse/JP-1449?focusedCommentId=467785#comment-467785):_

Email exchange -

Hi Deepashri

We could go through the functions and classes and get back to Dave.

A list of functions and classes would be a good place to start - maybe a recursive grep on "class " and "def " might be a way to get them, unless Dave has a neater way to extract them...

Anand
*______________________________________*
From: Deepashri Thatte
Sent: Thursday, June 4, 2020 11:18:02 AM
To: Anand Sivaramakrishnan; David Grumm
Cc: Rachel Cooper
Subject: Re: Question about the new code

Hi Dave,

Few additional functions that are not needed by the pipeline and I can quickly think of are listed below. There are definitely more functions that are not used and I will look at it.


implaneia_driver.py     simulate_data

utils.py                         amisim2mirage


Anand, how about you and I talk first to come up with a list of functions that are not needed by the pipeline? I will go over the modules and make a list before I talk to you. As you suggested we can then talk to Dave to discuss the flow of the algorithm.


Here are few more things in addition to what Anand mentioned. Please note that this is not a complete list.


-We are providing a pixel scale and not finding it. In the earlier version we provided a list of pixel scales listed as 'relpixelscales' and then found an optimal pixel scale.

-We are not using new_and_better_scaling_routine (I think this is nrm_model.improve_scaling in the pipeline).

-In ami_analyze.cfg we have a parameter called 'rotate' that we were providing. In the current version user has an option to find this value or provide a known value for 'rotate'.

-There are additional steps before we use fit_image. We find centroid of the data, find offset of the peak pixel from pixel center and use that offset in creating an analytical model at the same offset as the data.


Adding some more information to Anand's comment "implaneia uses simulated data that has mirage headers grafted onto the data.  There are some new functions associated with this operation.  These are called in the sample driver in the 'delivery' branch."

Pipeline can eliminate these steps since it will use _cal.fits file. The functions associated with this operation are

implaneia_driver.py     simulate_data

utils.py                         amisim2mirage



Thanks

Deepashri


*______________________________*
From: Anand Sivaramakrishnan
Sent: Thursday, June 4, 2020 9:26 AM
To: David Grumm; Deepashri Thatte
Cc: Rachel Cooper
Subject: Re: Question about the new code

Hi Dave

Deepashri might have complementary views, corrections or additions to this.  Would you be able to send us a list of the new functions?

In response to your question "If you have any general suggestions that might make my task easier I would appreciate it.",  it might help to describe the code changes between "ImPlaneIA" (aka implaneia or LG++) and our 2017 "original" (LG+) code.with a broad brush.

Our starting point for the new code remains the same as the old one: it supports different masks on different telescopes.     It also greatly simplifies use, with far fewer decisions made in deeper level functions, and far fewer in-line workarounds/assumptions made by default.  Hopefully this simplifies future use, debugging, and refactoring.

Some highlights of the changes are:

implaneia uses simulated data that has mirage headers grafted onto the data.  There are some new functions associated with this operation.  These are called in the sample driver in the 'delivery' branch.

implaneia uses a new class (Affine2d in nrm_analysis/misctools/utils.py) extensively.  This is used in the delivery branch.  Some of its methods are not used in the delivery branch.

implaneia has more functions dedicated to understandable and succinct output.  Some re used in the sample driver, some for development.

implaneia still has LG+'s  "Calibrate" class in nrm_core.py, but that is not used now for our processing, as it was fragile in the way it depended on particular input file structure.  I would like to remove that object entirely.  You could remove it in your pipeline code.

implaneia still has LG+'s  "BinaryAnalyze" class in nrm_core.py, but that is not used in the sample driver.  We might remove it from implaneia or split it out from nrm_core.py and rewrite it from scratch.   You could remove it in your pipeline code.


nrm_analysis/test directory:

- Some of these unit tests are usable unit tests, some work the routines but do not test any output.  These scripts were written when developing big sections (e.g. Affine2d).
- test_fringe_fitting.py seems to import a broken error propagation module.


Pipeline flow:

The driver in our 'delivery' branch of implaneia stops after writing text "observables" for a given 3-D mirage-header fits cube that are a stack of equal exposure-depth slices.  The IDT is coding simplified calibration drivers or functions/objects to average.  The tasks involved include those currently done by the pipeline's "ami_average" and the ami_(normalize?) to use the PSF calibrator observables to remove  instrument & telescope effects from the science target observables (i.e., perform interferometric 'calibration').  The new code uses an "ObservablesFromText class (which isn't yet placed neatly inside a module).  This development is still under way by three coders, and is a bit messy, with duplicated code.

implaneia now has the ability to output OIFITS (v2) files, which are fits binary tables that implement a published interferometric standard.  I hope we can output these fits files somehow in the pipeline, since these are typically the result of averaging and calibrating observables, yet contain essential coordinate system, baseline orientation, and time information.


Would you want to talk about this with a (proper or not) subset of us if it speeds up and/or simplifies your work?

Rachel Cooper uses implaneia and has written code for parts of it, so seeing this conversation might be useful - hence the cc.

Thanks
Anand


*______________________________________*
From: David Grumm
Sent: Thursday, June 4, 2020 2:15:36 AM
To: Deepashri Thatte
Cc: Anand Sivaramakrishnan; David Grumm
Subject: Question about the new code

Hi Deepashri,


To incorporate the new algorithm into the code running in our pipeline, I have run the tutorial and some of the notebooks you have provided to become familiar with the new code.  At this point I want to compare the functions in the new code to the functions in a copy I have of your "original" (2017) code, and using those differences, update the existing pipeline code. I just did a quick count of the functions in both your old and new code, and they have about 160 and 300 functions.  It may be that some of the functions you have in the new code were for development purposes. Do you know if the tests that you provided give an accurate coverage of all of the functions that would be used in the pipeline ?  If you have any general suggestions that might make my task easier I would appreciate it.


Thanks,

Dave

**Comment by stscijgbot-jp** [comment]
_Comment by [David Grumm](https://jira.stsci.edu/secure/ViewProfile.jspa?name=grumm) on [JIRA](https://jira.stsci.edu/browse/JP-1449?focusedCommentId=472131#comment-472131):_

1) Out of over 250 functions that are in the latest code base I received from Anand and Deepashri,  I have compiled a list of only 67 of those functions that will be used in the pipeline. As I've mentioned, these code updates are code updates to their earlier code, not to the code that has been running in the pipeline. I have started migrating them into the step code, refactoring them as needed.

2) As part of the ami_analyze step update, I have modified a nan-replacement routine from the coronography pipeline code and tested in within my local copy of the ami_analyze step code. The routine replaces nans in the SCI array with the median of the neighboring pixels having valid values. Although the data quality flag array is not currently used or output by the step, I am modifiyng the code to update the data quality flag for reset pixels, for possible later (local) use  in the step when a mask is needed for FFT calculations.
